Added ADDensity

tophmatthews opened the pull request:
(#12633)

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

tophmatthews committed:
Added ADDensity
(#12633)

tophmatthews commented:
@lindsayad When compiling ADDensity, I get two errors related to `_ad_zero` and `_ad_grad_zero`:

```
/Users/topher/projects/moose/modules/misc/src/materials/ADDensity.C:30:55: error: incompatible operand types ('const MooseArray<libMesh::Real>' and 'const MooseArray<ADReal>')
                : (this->template isCoupled("disp_r") ? adCoupledValue("disp_r") : _ad_zero)),

...

/Users/topher/projects/moose/modules/misc/src/materials/ADDensity.C:53:16: error: no matching member function for call to 'resize'
    _grad_disp.resize(3, &_ad_grad_zero);
```

Do you know why this is? I tried `this->template _zero`. both `adCoupledValue("disp_r")` and  `_ad_zero` should be `const MooseArray<ADReal>` no?

lindsayad commented:
`adCoupledValue` returns either a MooseArray of `Reals` or `ADReals` depending on the value of `compute_stage` (analogous for `adCoupledGradient`). This is to prevent derivative operations when computing the residual (saving a lot of unnecessary computational expense). The macros you are looking for are `adZero` and `adGradZero`. So try:
```c++
 (this->isCoupled("disp_r") ? adCoupledValue("disp_r") : adZero())),
```
and
```c++
_grad_disp.resize(3, &adGradZero());
```
Let me know if that doesn't work and I'll pull your PR

tophmatthews commented:
Yup that did it, thanks!

tophmatthews commented on modules/misc/src/materials/ADDensity.C:
```diff
@@ -0,0 +1,137 @@
+    // rho = rho0 / det(F)
+    // rho = rho0 / det(grad(u) + 1)
+
+    const auto Axx = (*_grad_disp[0])[_qp](0) + 1.0;
```
@lindsayad  Well, using `adZero` and `adGradZero` worked, but now I get a seg fault within these lines during the first linear iteration. It runs initially, the (0) non-linear iteration is calculated `0 Nonlinear |R| = 1.414214e+02`, then it `Access out of bounds in MooseArray (i: 0 size: 0)`. I'm assuming this is due to the complexities added by the AD which adds another vector dimension. This happens when coupling the displacements altogether `displacements = 'disp_x disp_y disp_z'`, or individually. I'm not sure why it works initially, only to have it fail on the first linear iteration. Any ideas?

lindsayad replied:
I don't see the reason off the top of my head, but I could probably figure it out pretty quickly if you give me an input file

tophmatthews replied:
I added tests to `modules/combined/test/tests/elastic_patch`. Interestingly, `ad_elastic_patch_rspherical.i` works, but `ad_elastic_patch_plane_strain.i` and `ad_elastic_patch_rz.i` don't.

tophmatthews replied:
It's something to do with the differences between RESIDUAL and JACOBIAN template specifications. When I do 
```
    std::cout << "stage: " << compute_stage << "  (*_grad_disp[0])[_qp](0): " << (*_grad_disp[0])[_qp](0) << std::endl;
    std::cout << "stage: " << compute_stage << "  (*_grad_disp[0])[_qp](1): " << (*_grad_disp[0])[_qp](1) << std::endl;
    std::cout << "stage: " << compute_stage << "  (*_grad_disp[0])[_qp](2): " << (*_grad_disp[0])[_qp](2) << std::endl;
    std::cout << "stage: " << compute_stage << "  (*_grad_disp[1])[_qp](0): " << (*_grad_disp[1])[_qp](0) << std::endl;
    std::cout << "stage: " << compute_stage << "  (*_grad_disp[1])[_qp](1): " << (*_grad_disp[1])[_qp](1) << std::endl;
    std::cout << "stage: " << compute_stage << "  (*_grad_disp[1])[_qp](2): " << (*_grad_disp[1])[_qp](2) << std::endl;
```
I get
```
...
stage: 0  (*_grad_disp[0])[_qp](0): 0
stage: 0  (*_grad_disp[0])[_qp](1): 0
stage: 0  (*_grad_disp[0])[_qp](2): 0
stage: 0  (*_grad_disp[1])[_qp](0): 0
stage: 0  (*_grad_disp[1])[_qp](1): 0
stage: 0  (*_grad_disp[1])[_qp](2): 0
 0 Nonlinear |R| = 2.012434e+01
stage: 1  (*_grad_disp[0])[_qp](0): (0,{-3.28615,3.72641,0.88052,-1.32078,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0})
stage: 1  (*_grad_disp[0])[_qp](1): (0,{-32.8615,-12.7359,8.8052,36.7922,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0})
stage: 1  (*_grad_disp[0])[_qp](2): (0,{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0})
stage: 1  (*_grad_disp[1])[_qp](0): (Segmentation fault: 11
```

tophmatthews replied:
I can get RESIDUAL to run okay and match with tests. It has to do with the grad terms in the jacobian calculations... help! :)

tophmatthews replied:
@lindsayad Yeah, something to do with the jacobian compute stage again, but only in RZ. You can reproduce the error with `modules/combined/test/tests/elastic_patch/ad_elastic_patch_rz_nonlinear.i`

tophmatthews replied:
Should do it!

tophmatthews committed:
Added tests
(#12633)

tophmatthews commented:
FYI, I added duplicative tests to the elastic patch tests in `combined/test/tests/elastic_patch`. I simplified the input files that were already there by removing the temperature kernels and boundary conditions since nothing in the input files are temperature dependent. I had to regold everything since I needed to add `outputs = all` to the `Density` and `ADDensity` material properties in order to make sure I was checking against them. @dschwen you may want to add to these tests when you get the AD tensor mechanics pieces working. They seem like good verification tests. Also, once everything with AD is replaced, you can add tests for the jacobian.

tophmatthews commented:
I also removed the `max_parallel = 1` lines from `tests` since it was added when a serial linear solver was used.

tophmatthews committed:
Added more tests for ADDensity
(#12633)

tophmatthews committed:
Fixes for ADDensity residual calc. Still needs jacobian
(#12633)

tophmatthews commented:
Not ready to merge, I just want to make sure the tests are passing

lindsayad commented:
I have a fix. I'll push up a patch to your PR soon

On Mon, Jan 7, 2019 at 8:12 PM MOOSE Build <notifications@github.com> wrote:

> Job Documentation <https://civet.inl.gov/job/271381/> on e3fc600
> <https://github.com/idaholab/moose/commit/e3fc6004b92d34cafb0b533b4d65823e5c7285cf>
> wanted to post the following:
>
> View the site here <http://mooseframework.inl.gov/docs/PRs/12649/site>
>
> This comment will be updated on new commits.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/12649#issuecomment-452160421>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJxgcK_WwQ2WV8Fb47tWAR-5sLqdkJq-ks5vBAylgaJpZM4ZtgoW>
> .
>

lindsayad committed:
Ensure AD calculations are always done when needed

lindsayad commented on modules/misc/src/materials/ADDensity.C:
```diff
@@ -0,0 +1,143 @@
+ADDensity<compute_stage>::ADDensity(const InputParameters & parameters)
+  : ADMaterial<compute_stage>(parameters),
+    _is_coupled(true),
+    _disp_r(this->template isCoupled("displacements")
```
The `template` keyword is only necessary when you are explicitly specifying a template argument, e.g. with angle brackets `<>`

lindsayad replied:
I'm going to push up another patch which adds a few members to `usingMaterialMembers` so that the `this->` syntax is not necessary at all.

tophmatthews replied:
this needs q_point?

tophmatthews replied:
@lindsayad adding `using ADKernel<compute_stage>::_q_point; ` to `ADKernel.h` doesn't capture this...

tophmatthews replied:
whoops, should go in `ADMaterial.h`, but has compiling error...

lindsayad committed:
Add more members to using*Members macros; eliminate need for this-> syntax

tophmatthews committed:
Reimplemented compute_stage template

tophmatthews committed:
Added q_point to avoid this-> notation
(#12633)

tophmatthews committed:
Rename original density to initial density
(#12633)

lindsayad committed:
Use more fool-proof method to determine when we're computing the Jacobian

tophmatthews committed:
Test cleanup
(#12633)

tophmatthews commented:
@lindsayad this should be good to go for full review

lindsayad commented on modules/misc/src/materials/ADDensity.C:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  // old deprecated parameters
+  else if (is_x_coupled || is_r_coupled)
```
Since we're creating a new class and we have a clean slate here, I think we should either eliminate code that is "deprecated" or not call it deprecated

tophmatthews replied:
I wanted to remove, but I know it's been pushed in the past to maintain all previous features. I'll rip this out to just use displacements.

lindsayad commented on modules/misc/src/materials/ADDensity.C:
```diff
@@ -0,0 +1,137 @@
+  else
+  {
+    _is_coupled = false;
+    // TODO: We should deprecate this case and have the user use a GenericConstantMaterial for this
```
Same statement as above

lindsayad commented on modules/misc/src/materials/ADDensity.C:
```diff
@@ -0,0 +1,137 @@
+    // rho = rho0 / det(grad(u) + 1)
+
+    const auto Axx = (*_grad_disp[0])[_qp](0) + 1.0;
+    const auto Axy = (*_grad_disp[0])[_qp](1);
```
Let's use `const auto &` to prevent construction of another dual number. Same goes for Axz, Ayx, Ayz, Azx, Azy

lindsayad reviewed (changes requested):
So I noticed that there are no tests using the `PetscJacobianTester`. If we did, I am confident they would fail because in order to use the derivative information coming from an `ADMaterial` you must use an `ADKernel`. What kernel/material actually uses the `density` property? Is this PR meant to enable future PRs which will actually exercise the derivative information coming from `ADDensity`?

tophmatthews commented:
This needs another kernel for usage, namely `ADHeatConductionTimeDerivative` #12648 which needs `ADTimeDerivative`. I was waiting of a kernel usage before doing a `PetscJacobianTester`. This is similar to some of the other tests I've done which require @dschwen tensor mechanics implementation before the entire input deck could be AD.

lindsayad commented:
> This needs another kernel for usage, namely `ADHeatConductionTimeDerivative` #12648 which needs `ADTimeDerivative`. I was waiting of a kernel usage before doing a `PetscJacobianTester`. This is similar to some of the other tests I've done which require @dschwen tensor mechanics implementation before the entire input deck could be AD.

Ok good, just wanted to get clear on the map forward for this class.

tophmatthews committed:
Removed old ways of doing things
(#12633)

moosebuild commented:
Job [Private App tests](https://civet.inl.gov/job/272554/) on aab5194 : invalidated by @permcody

Xgels server was down

moosebuild commented:
Job [Private App tests PETSc alt](https://civet.inl.gov/job/272557/) on aab5194 : invalidated by @permcody

Xgels server was down

dschwen commented on framework/include/materials/ADMaterial.h:
```diff
@@ -17,7 +17,19 @@
+  using ADMaterial<compute_stage>::coupled;                                                        \
+  using ADMaterial<compute_stage>::coupledComponents;                                              \
+  using ADMaterial<compute_stage>::isParamValid;                                                   \
+  using ADMaterial<compute_stage>::getParam;                                                       \
```
This won't work. See discussion in #12657

tophmatthews replied:
I'm not even using it, so I'll drop it from this PR

dschwen commented:
Whats with the "Illegal Instruction" crashing tests?

tophmatthews commented:
> Whats with the "Illegal Instruction" crashing tests?

!!!!! no idea

moosebuild commented:
Job [Test Intel](https://civet.inl.gov/job/272895/) on 3f40443 : invalidated by @tophmatthews

moosebuild commented:
Job [Pthreads Test](https://civet.inl.gov/job/272887/) on 3f40443 : invalidated by @tophmatthews

tophmatthews commented:
@lindsayad okay good to go for review

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/274331/) on ef70a6e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/12649/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/12649/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1abf649307c970e593233366e909053126cb5402`

tophmatthews committed:
Removed getParam
(#12633)
